### PR TITLE
PluginCapability update for external-attacher

### PR DESCRIPTION
**Pull request by NickrenREN**
PluginCapability update for external-attacher
/cc @saad-ali @jsafrane

### CODE REVIEW

File: cmd/csi-attacher/main.go [hunk 1]
```diff
@@ -109,7 +109,12 @@ func main() {
 			glog.Error(err.Error())
 			os.Exit(1)
 		}
-		if !supportsAttach {
+		supportsService, err := csiConn.SupportsPluginControllerService(ctx)
+		if err != nil {
+			glog.Error(err.Error())
+			os.Exit(1)
+		}
+		if !supportsAttach || !supportsService {
```
Comment (jsafrane) [inline cmd/csi-attacher/main.go]: I think you should check for `SupportsPluginControllerService` first and only when the driver supports controller service you can invoke controller methods such as `ControllerGetCapabilities` used in `SupportsControllerPublish`.

File: pkg/connection/connection.go [hunk 2]
```diff
@@ -164,6 +168,30 @@ func (c *csiConnection) SupportsControllerPublish(ctx context.Context) (bool, er
 	return false, nil
 }
 
+func (c *csiConnection) SupportsPluginControllerService(ctx context.Context) (bool, error) {
```
Comment (jsafrane) [inline pkg/connection/connection.go]: please add unit test for this function.

File: cmd/csi-attacher/main.go [hunk 3]
```diff
@@ -103,21 +103,28 @@ func main() {
...
-			vaLister := factory.Storage().V1alpha1().VolumeAttachments().Lister()
-			handler = controller.NewCSIHandler(clientset, attacher, csiConn, pvLister, nodeLister, vaLister)
-			glog.V(2).Infof("CSI driver supports ControllerPublishUnpublish, using real CSI handler")
+			// Find out if the driver supports attach/detach.
+			supportsAttach, err := csiConn.SupportsControllerPublish(ctx)
+			if err != nil {
+				glog.Error(err.Error())
+				os.Exit(1)
+			}
+			if supportsAttach {
```
Comment (jsafrane) [inline cmd/csi-attacher/main.go]: Add `else { /* fall back to trivial handler */ }` to this `if`.

Comment (NickrenREN) [reply cmd/csi-attacher/main.go]: oh, sorry, i misunderstand this check, will update right now

Comment (NickrenREN) [reply cmd/csi-attacher/main.go]: updated, PTAL @jsafrane , thanks

File: pkg/connection/connection.go (hunk 2 above)
Comment (NickrenREN) [reply pkg/connection/connection.go]: done

**Comment by jsafrane** [comment]
Thanks for the PR!

**Comment by saad-ali** [comment]
Thanks @NickrenREN!

**Pull request merged**